File: http/tests/fw/api/types/validation/result.py
from __future__ import annotations

from tests.fw.util import AttrDict, AttrHookDef
from .val_activation_blocked import ValActivationBlockedFail
from .val_capital_module import ValCapModuleFail
from .val_charge_group import ValChargeGroupFail
from .val_charge_size import ValChargeSizeFail
from .val_charge_volume import ValChargeVolumeFail
from .val_drone_group import ValDroneGroupFail
from .val_effect_stopper import ValEffectStopperFail
from .val_fighter_squad_size import ValFighterSquadSizeFail
from .val_item_kind import ValItemKindFail
from .val_item_vs_ship_kind import ValItemVsShipKindFail
from .val_max_group import ValMaxGroupFail
from .val_max_type import ValMaxTypeFail
from .val_module_state import ValModuleStateFail
from .val_not_loaded_item import ValNotLoadedItemFail
from .val_overload_skill import ValOverloadSkillFail
from .val_proj_immunity import ValProjImmunityFail
from .val_resources import ValResourceFail
from .val_rig_size import ValRigSizeFail
from .val_sec_zone import ValSecZoneFail
from .val_ship_limit import ValShipLimitFail
from .val_ship_stance import ValShipStanceFail
from .val_skill_reqs import ValSrqFail
from .val_slot_count import ValSlotCountFail
from .val_slot_index import ValSlotIndexFail
from .val_unusable_resource import ValUnusableResFail
from .val_unusable_slot import ValUnusableSlotFail


class SolValResult(AttrDict):

    def __init__(self, *, data: dict) -> None:
        super().__init__(data=data, hooks={'details': AttrHookDef(func=lambda d: SolValDetails(data=d))})

    @property
    def fits(self) -> dict[str, FitValDetails]:
        details = getattr(self, 'details', AttrDict(data={}))
        return getattr(details, 'fits', {})


class SolValDetails(AttrDict):

    def __init__(self, *, data: dict) -> None:
        super().__init__(data=data, hooks={
            'fits': AttrHookDef(func=lambda d: {k: FitValDetails(data=v) for k, v in d.items()}),
            'not_loaded_item': AttrHookDef(func=lambda d: ValNotLoadedItemFail(data=d))})


class FitValResult(AttrDict):

    def __init__(self, *, data: dict) -> None:
        super().__init__(data=data, hooks={'details': AttrHookDef(func=lambda d: FitValDetails(data=d))})


class FitValDetails(AttrDict):

    def __init__(self, *, data: dict) -> None:
        super().__init__(data=data, hooks={
            'cpu': AttrHookDef(func=lambda d: ValResourceFail(data=d)),
            'powergrid': AttrHookDef(func=lambda d: ValResourceFail(data=d)),
            'calibration': AttrHookDef(func=lambda d: ValResourceFail(data=d)),
            'drone_bay_volume': AttrHookDef(func=lambda d: ValResourceFail(data=d)),
            'drone_bandwidth': AttrHookDef(func=lambda d: ValResourceFail(data=d)),
            'fighter_bay_volume': AttrHookDef(func=lambda d: ValResourceFail(data=d)),
            'rig_slot_count': AttrHookDef(func=lambda d: ValSlotCountFail(data=d)),
            'service_slot_count': AttrHookDef(func=lambda d: ValSlotCountFail(data=d)),
            'subsystem_slot_count': AttrHookDef(func=lambda d: ValSlotCountFail(data=d)),
            'launched_drone_count': AttrHookDef(func=lambda d: ValSlotCountFail(data=d)),
            'launched_fighter_count': AttrHookDef(func=lambda d: ValSlotCountFail(data=d)),
            'launched_light_fighter_count': AttrHookDef(func=lambda d: ValSlotCountFail(data=d)),
            'launched_heavy_fighter_count': AttrHookDef(func=lambda d: ValSlotCountFail(data=d)),
            'launched_support_fighter_count': AttrHookDef(func=lambda d: ValSlotCountFail(data=d)),
            'launched_st_light_fighter_count': AttrHookDef(func=lambda d: ValSlotCountFail(data=d)),
            'launched_st_heavy_fighter_count': AttrHookDef(func=lambda d: ValSlotCountFail(data=d)),
            'launched_st_support_fighter_count': AttrHookDef(func=lambda d: ValSlotCountFail(data=d)),
            'turret_slot_count': AttrHookDef(func=lambda d: ValSlotCountFail(data=d)),
            'launcher_slot_count': AttrHookDef(func=lambda d: ValSlotCountFail(data=d)),
            'high_slot_count': AttrHookDef(func=lambda d: ValSlotCountFail(data=d)),
            'mid_slot_count': AttrHookDef(func=lambda d: ValSlotCountFail(data=d)),
            'low_slot_count': AttrHookDef(func=lambda d: ValSlotCountFail(data=d)),
            'implant_slot_index': AttrHookDef(func=lambda d: ValSlotIndexFail(data=d)),
            'booster_slot_index': AttrHookDef(func=lambda d: ValSlotIndexFail(data=d)),
            'subsystem_slot_index': AttrHookDef(func=lambda d: ValSlotIndexFail(data=d)),
            'ship_limit': AttrHookDef(func=lambda d: ValShipLimitFail(data=d)),
            'max_group_fitted': AttrHookDef(func=lambda d: ValMaxGroupFail(data=d)),
            'max_group_online': AttrHookDef(func=lambda d: ValMaxGroupFail(data=d)),
            'max_group_active': AttrHookDef(func=lambda d: ValMaxGroupFail(data=d)),
            'rig_size': AttrHookDef(func=lambda d: ValRigSizeFail(data=d)),
            'skill_reqs': AttrHookDef(func=lambda d: ValSrqFail(data=d)),
            'charge_group': AttrHookDef(func=lambda d: ValChargeGroupFail(data=d)),
            'charge_size': AttrHookDef(func=lambda d: ValChargeSizeFail(data=d)),
            'charge_volume': AttrHookDef(func=lambda d: ValChargeVolumeFail(data=d)),
            'capital_module': AttrHookDef(func=lambda d: ValCapModuleFail(data=d)),
            'not_loaded_item': AttrHookDef(func=lambda d: ValNotLoadedItemFail(data=d)),
            'module_state': AttrHookDef(func=lambda d: ValModuleStateFail(data=d)),
            'item_kind': AttrHookDef(func=lambda d: ValItemKindFail(data=d)),
            'drone_group': AttrHookDef(func=lambda d: ValDroneGroupFail(data=d)),
            'fighter_squad_size': AttrHookDef(func=lambda d: ValFighterSquadSizeFail(data=d)),
            'unlaunchable_drone_slot': AttrHookDef(func=lambda d: ValUnusableSlotFail(data=d)),
            'unlaunchable_drone_bandwidth': AttrHookDef(func=lambda d: ValUnusableResFail(data=d)),
            'unlaunchable_fighter': AttrHookDef(func=lambda d: ValUnusableSlotFail(data=d)),
            'unlaunchable_light_fighter': AttrHookDef(func=lambda d: ValUnusableSlotFail(data=d)),
            'unlaunchable_heavy_fighter': AttrHookDef(func=lambda d: ValUnusableSlotFail(data=d)),
            'unlaunchable_support_fighter': AttrHookDef(func=lambda d: ValUnusableSlotFail(data=d)),
            'unlaunchable_st_light_fighter': AttrHookDef(func=lambda d: ValUnusableSlotFail(data=d)),
            'unlaunchable_st_heavy_fighter': AttrHookDef(func=lambda d: ValUnusableSlotFail(data=d)),
            'unlaunchable_st_support_fighter': AttrHookDef(func=lambda d: ValUnusableSlotFail(data=d)),
            'ship_stance': AttrHookDef(func=lambda d: ValShipStanceFail(data=d)),
            'overload_skill': AttrHookDef(func=lambda d: ValOverloadSkillFail(data=d)),
            'max_type_fitted': AttrHookDef(func=lambda d: ValMaxTypeFail(data=d)),
            'sec_zone_fitted': AttrHookDef(func=lambda d: ValSecZoneFail(data=d)),
            'sec_zone_online': AttrHookDef(func=lambda d: ValSecZoneFail(data=d)),
            'sec_zone_active': AttrHookDef(func=lambda d: ValSecZoneFail(data=d)),
            'sec_zone_unonlineable': AttrHookDef(func=lambda d: ValSecZoneFail(data=d)),
            'sec_zone_unactivable': AttrHookDef(func=lambda d: ValSecZoneFail(data=d)),
            'activation_blocked': AttrHookDef(func=lambda d: ValActivationBlockedFail(data=d)),
            'item_vs_ship_kind': AttrHookDef(func=lambda d: ValItemVsShipKindFail(data=d)),
            'effect_stopper': AttrHookDef(func=lambda d: ValEffectStopperFail(data=d)),
            'assist_immunity': AttrHookDef(func=lambda d: ValProjImmunityFail(data=d)),
            'offense_immunity': AttrHookDef(func=lambda d: ValProjImmunityFail(data=d)),
            'resist_immunity': AttrHookDef(func=lambda d: ValProjImmunityFail(data=d))})

File: core/src/sol/svc/vast/result.rs
use std::collections::HashMap;

use crate::sol::{
    FitId,
    svc::vast::{
        ValActivationBlockedFail, ValCapitalModFail, ValChargeGroupFail, ValChargeSizeFail, ValChargeVolumeFail,
        ValDroneGroupFail, ValEffectStopperFail, ValFighterSquadSizeFail, ValItemKindFail, ValItemVsShipKindFail,
        ValMaxGroupFail, ValMaxTypeFail, ValModuleStateFail, ValNotLoadedItemFail, ValOverloadSkillFail,
        ValProjImmunityFail, ValResFail, ValRigSizeFail, ValSecZoneFail, ValShipLimitFail, ValShipStanceFail,
        ValSlotCountFail, ValSlotIndexFail, ValSrqFail, ValUnusableResFail, ValUnusableSlotFail,
    },
};

/// Validation result for a solar system.
///
/// Contains per-fit failures, and failures for items not belonging to any fit.
pub struct ValResultSol {
    pub fits: HashMap<FitId, ValResultFit>,
    /// Not loaded stand-alone items.
    pub not_loaded_item: Option<ValNotLoadedItemFail>,
}
impl ValResultSol {
    pub(in crate::sol::svc::vast) fn new() -> Self {
        Self {
            fits: HashMap::new(),
            not_loaded_item: None,
        }
    }
    pub fn all_passed(&self) -> bool {
        self.fits.is_empty() && self.not_loaded_item.is_none()
    }
}

/// Validation result for a fit.
pub struct ValResultFit {
    // Generic
    pub not_loaded_item: Option<ValNotLoadedItemFail>,
    pub item_kind: Option<ValItemKindFail>,
    pub skill_reqs: Option<ValSrqFail>,
    // Implants/boosters
    pub implant_slot_index: Option<ValSlotIndexFail>,
    pub booster_slot_index: Option<ValSlotIndexFail>,
    // Shared between mod-alike items
    pub cpu: Option<ValResFail>,
    pub powergrid: Option<ValResFail>,
    pub ship_limit: Option<ValShipLimitFail>,
    pub max_group_fitted: Option<ValMaxGroupFail>,
    pub max_group_online: Option<ValMaxGroupFail>,
    pub max_group_active: Option<ValMaxGroupFail>,
    pub max_type_fitted: Option<ValMaxTypeFail>,
    pub item_vs_ship_kind: Option<ValItemVsShipKindFail>,
    // Modules
    pub high_slot_count: Option<ValSlotCountFail>,
    pub mid_slot_count: Option<ValSlotCountFail>,
    pub low_slot_count: Option<ValSlotCountFail>,
    pub turret_slot_count: Option<ValSlotCountFail>,
    pub launcher_slot_count: Option<ValSlotCountFail>,
    pub module_state: Option<ValModuleStateFail>,
    pub capital_module: Option<ValCapitalModFail>,
    pub overload_skill: Option<ValOverloadSkillFail>,
    // Charges
    pub charge_group: Option<ValChargeGroupFail>,
    pub charge_size: Option<ValChargeSizeFail>,
    pub charge_volume: Option<ValChargeVolumeFail>,
    // Rigs
    pub rig_slot_count: Option<ValSlotCountFail>,
    pub calibration: Option<ValResFail>,
    pub rig_size: Option<ValRigSizeFail>,
    // Services
    pub service_slot_count: Option<ValSlotCountFail>,
    // T3 subsystems/stances
    pub subsystem_slot_count: Option<ValSlotCountFail>,
    pub subsystem_slot_index: Option<ValSlotIndexFail>,
    pub ship_stance: Option<ValShipStanceFail>,
    // Drones
    pub drone_bay_volume: Option<ValResFail>,
    pub launched_drone_count: Option<ValSlotCountFail>,
    pub drone_bandwidth: Option<ValResFail>,
    pub unlaunchable_drone_slot: Option<ValUnusableSlotFail>,
    pub unlaunchable_drone_bandwidth: Option<ValUnusableResFail>,
    pub drone_group: Option<ValDroneGroupFail>,
    // Fighters
    pub fighter_bay_volume: Option<ValResFail>,
    pub launched_fighter_count: Option<ValSlotCountFail>,
    pub launched_light_fighter_count: Option<ValSlotCountFail>,
    pub launched_heavy_fighter_count: Option<ValSlotCountFail>,
    pub launched_support_fighter_count: Option<ValSlotCountFail>,
    pub launched_st_light_fighter_count: Option<ValSlotCountFail>,
    pub launched_st_heavy_fighter_count: Option<ValSlotCountFail>,
    pub launched_st_support_fighter_count: Option<ValSlotCountFail>,
    pub unlaunchable_fighter: Option<ValUnusableSlotFail>,
    pub unlaunchable_light_fighter: Option<ValUnusableSlotFail>,
    pub unlaunchable_heavy_fighter: Option<ValUnusableSlotFail>,
    pub unlaunchable_support_fighter: Option<ValUnusableSlotFail>,
    pub unlaunchable_st_light_fighter: Option<ValUnusableSlotFail>,
    pub unlaunchable_st_heavy_fighter: Option<ValUnusableSlotFail>,
    pub unlaunchable_st_support_fighter: Option<ValUnusableSlotFail>,
    pub fighter_squad_size: Option<ValFighterSquadSizeFail>,
    // Projection, destination side
    pub activation_blocked: Option<ValActivationBlockedFail>,
    pub effect_stopper: Option<ValEffectStopperFail>,
    // Projection, source side
    pub assist_immunity: Option<ValProjImmunityFail>,
    pub offense_immunity: Option<ValProjImmunityFail>,
    pub resist_immunity: Option<ValProjImmunityFail>,
    // Sec zone
    pub sec_zone_fitted: Option<ValSecZoneFail>,
    pub sec_zone_online: Option<ValSecZoneFail>,
    pub sec_zone_active: Option<ValSecZoneFail>,
    pub sec_zone_unonlineable: Option<ValSecZoneFail>,
    pub sec_zone_unactivable: Option<ValSecZoneFail>,
}
impl ValResultFit {
    pub(in crate::sol::svc::vast) fn new() -> Self {
        Self {
            // Generic
            not_loaded_item: None,
            item_kind: None,
            skill_reqs: None,
            // Implants/boosters
            implant_slot_index: None,
            booster_slot_index: None,
            // Shared between mod-alike items
            cpu: None,
            powergrid: None,
            ship_limit: None,
            max_group_fitted: None,
            max_group_online: None,
            max_group_active: None,
            max_type_fitted: None,
            item_vs_ship_kind: None,
            // Modules
            high_slot_count: None,
            mid_slot_count: None,
            low_slot_count: None,
            turret_slot_count: None,
            launcher_slot_count: None,
            module_state: None,
            capital_module: None,
            overload_skill: None,
            // Charges
            charge_group: None,
            charge_size: None,
            charge_volume: None,
            // Rigs
            rig_slot_count: None,
            calibration: None,
            rig_size: None,
            // Services
            service_slot_count: None,
            // T3 subsystems/stances
            subsystem_slot_count: None,
            subsystem_slot_index: None,
            ship_stance: None,
            // Drones
            drone_bay_volume: None,
            launched_drone_count: None,
            drone_bandwidth: None,
            unlaunchable_drone_slot: None,
            unlaunchable_drone_bandwidth: None,
            drone_group: None,
            // Fighters
            fighter_bay_volume: None,
            launched_fighter_count: None,
            launched_light_fighter_count: None,
            launched_heavy_fighter_count: None,
            launched_support_fighter_count: None,
            launched_st_light_fighter_count: None,
            launched_st_heavy_fighter_count: None,
            launched_st_support_fighter_count: None,
            unlaunchable_fighter: None,
            unlaunchable_light_fighter: None,
            unlaunchable_heavy_fighter: None,
            unlaunchable_support_fighter: None,
            unlaunchable_st_light_fighter: None,
            unlaunchable_st_heavy_fighter: None,
            unlaunchable_st_support_fighter: None,
            fighter_squad_size: None,
            // Projection, destination side
            activation_blocked: None,
            effect_stopper: None,
            // Projection, source side
            assist_immunity: None,
            offense_immunity: None,
            resist_immunity: None,
            // Sec zone
            sec_zone_fitted: None,
            sec_zone_online: None,
            sec_zone_active: None,
            sec_zone_unonlineable: None,
            sec_zone_unactivable: None,
        }
    }
    pub fn all_passed(&self) -> bool {
        // Generic
        self.not_loaded_item.is_none()
            && self.item_kind.is_none()
            && self.skill_reqs.is_none()
            // Implants/boosters
            && self.implant_slot_index.is_none()
            && self.booster_slot_index.is_none()
            // Shared between mod-alike items
            && self.cpu.is_none()
            && self.powergrid.is_none()
            && self.ship_limit.is_none()
            && self.max_group_fitted.is_none()
            && self.max_group_online.is_none()
            && self.max_group_active.is_none()
            && self.max_type_fitted.is_none()
            && self.item_vs_ship_kind.is_none()
            // Modules
            && self.high_slot_count.is_none()
            && self.mid_slot_count.is_none()
            && self.low_slot_count.is_none()
            && self.turret_slot_count.is_none()
            && self.launcher_slot_count.is_none()
            && self.module_state.is_none()
            && self.capital_module.is_none()
            && self.overload_skill.is_none()
            // Charges
            && self.charge_group.is_none()
            && self.charge_size.is_none()
            && self.charge_volume.is_none()
            // Rigs
            && self.rig_slot_count.is_none()
            && self.calibration.is_none()
            && self.rig_size.is_none()
            // Services
            && self.service_slot_count.is_none()
            // T3 subsystems/stances
            && self.subsystem_slot_count.is_none()
            && self.subsystem_slot_index.is_none()
            && self.ship_stance.is_none()
            // Drones
            && self.drone_bay_volume.is_none()
            && self.launched_drone_count.is_none()
            && self.drone_bandwidth.is_none()
            && self.unlaunchable_drone_slot.is_none()
            && self.unlaunchable_drone_bandwidth.is_none()
            && self.drone_group.is_none()
            // Fighters
            && self.fighter_bay_volume.is_none()
            && self.launched_fighter_count.is_none()
            && self.launched_light_fighter_count.is_none()
            && self.launched_heavy_fighter_count.is_none()
            && self.launched_support_fighter_count.is_none()
            && self.launched_st_light_fighter_count.is_none()
            && self.launched_st_heavy_fighter_count.is_none()
            && self.launched_st_support_fighter_count.is_none()
            && self.unlaunchable_fighter.is_none()
            && self.unlaunchable_light_fighter.is_none()
            && self.unlaunchable_heavy_fighter.is_none()
            && self.unlaunchable_support_fighter.is_none()
            && self.unlaunchable_st_light_fighter.is_none()
            && self.unlaunchable_st_heavy_fighter.is_none()
            && self.unlaunchable_st_support_fighter.is_none()
            && self.fighter_squad_size.is_none()
            // Projection, destination side
            && self.activation_blocked.is_none()
            && self.effect_stopper.is_none()
            // Projection, source side
            && self.assist_immunity.is_none()
            && self.offense_immunity.is_none()
            && self.resist_immunity.is_none()
            // Sec zone
            && self.sec_zone_fitted.is_none()
            && self.sec_zone_online.is_none()
            && self.sec_zone_active.is_none()
            && self.sec_zone_unonlineable.is_none()
            && self.sec_zone_unactivable.is_none()
    }
}
